### PR TITLE
Shaharsol/eco 1515/incoming transfer

### PR DESCRIPTION
**Pull request by shaharsol**
This contains the _incoming_ side of the transfer mechanism. Pay special attention to modifications in services/payments.ts and internal/services.ts

### CODE REVIEW

File: scripts/src/client.ts [hunk 1]
```diff
@@ -417,6 +438,24 @@ export class Client {
...
+		try {
+			const res = await this.requests
+				.request(`/v2/transfers/incoming/orders`, {
+						wallet_address: walletAddress,
+						app_id: appId,
+						title,
+						description,
+						memo,
+				 })
+				.post<OpenOrder>();
```
Comment (doodyparizada) [inline scripts/src/client.ts]: Doesn't this return a regular Order?

File: scripts/src/internal/services.ts [hunk 2]
```diff
@@ -101,7 +102,12 @@ async function getPaymentJWT(order: db.Order, appId: string, user: User): Promis
 }
 
 export async function paymentComplete(payment: CompletedPayment) {
-	const order = await db.Order.getOne({ orderId: payment.id });
+	// if I can find the "paytment.id" in the redis, it means its a memo and the value is the real order id
+	const redis = getRedisClient();
+	const incomingOrderId = await redis.async.get("transfer_" + payment.id);
```
Comment (doodyparizada) [inline scripts/src/internal/services.ts]: please use a prefix with colon and use string substitution: 
```
.get(`transfer:${payment.id}`)
```

File: scripts/src/internal/services.ts [hunk 3]
```diff
@@ -178,8 +189,14 @@ export async function paymentComplete(payment: CompletedPayment) {
 	const prevStatus = order.status;
 	const prevStatusDate = order.currentStatusDate;
 	order.setStatus("completed");
+
 	await order.save();
 
+	// if it was an incoming transfer, clear the cache so we wont process again
+	if (incomingOrderId) {
+		await redis.async.del(payment.id);
```
Comment (doodyparizada) [inline scripts/src/internal/services.ts]: I think you might be able to skip this if you use `setex` to set this key.

**Review by doodyparizada — CHANGES_REQUESTED** [review]
Haven't gone through the whole thing but I have some comments

File: scripts/src/internal/services.ts (hunk 3 above)
Comment (shaharsol) [reply scripts/src/internal/services.ts]: I can, however it still won't invalidate the cache once the hook triggers, and won't protect from a 2nd trigger, so even if I use setex I think we should manually invalidate, don't you think?

File: scripts/src/client.ts (hunk 1 above)
Comment (shaharsol) [reply scripts/src/client.ts]: Right, it was returning OpenOrder but then I changed it to create the order as pending. Done.

File: scripts/src/internal/services.ts (hunk 2 above)
Comment (shaharsol) [reply scripts/src/internal/services.ts]: done

File: scripts/src/mock_client.ts [hunk 4]
```diff
@@ -1802,11 +1804,209 @@ async function checkOutgoingTransferOrder(){
...
+	console.log("===================================== checkIncomingTransferOrder =====================================");
+
+	const userId = generateId();
+	const deviceId = generateId();
+	const appClient = new SampleAppClient(SMPL_APP_CONFIG.jwtAddress);
+	const jwt = await appClient.getRegisterJWT(userId, deviceId);
+	const client = await MarketplaceClient.create({ jwt });
+	await client.updateWallet(SMPL_APP_CONFIG.keypair.publicKey());
+
+	// const receiverWalletAddress = `wallet-${ generateRandomString({ prefix: userId, length: 56 }) }`;
```
Comment (doodyparizada) [inline scripts/src/mock_client.ts]: delete comment

File: scripts/src/mock_client.ts [hunk 5]
```diff
@@ -1802,11 +1804,209 @@ async function checkOutgoingTransferOrder(){
...
+
+	console.log(`completion date: ${ order.completion_date }`);
+
+	// check order on blockchain
+	const apayment = (await retry(() => senderClient.findKinPayment(order.id), payment => !!payment, "failed to find payment on blockchain"))!;
+
+	console.log(`payment on blockchain:`, apayment);
+	// ============ end poll section ============
+
+	const util = require("util");
```
Comment (doodyparizada) [inline scripts/src/mock_client.ts]: delete this

File: scripts/src/mock_client.ts [hunk 6]
```diff
@@ -1802,11 +1804,209 @@ async function checkOutgoingTransferOrder(){
...
+
+	const incomingOrder = await receiverClient.createIncomingTransferOrder(senderKeys.publicKey(), "sender-app", "mock client title", "mock client description", memo);
+	expect(incomingOrder).toMatchObject({ title: "mock client title" });
+
+	const outgoingOrder = await senderClient.createOutgoingTransferOrder(receiverKeys.publicKey(), "receiver-app", "mock client title", "mock client description", memo, amount);
+	expect(outgoingOrder).toMatchObject({ title: "mock client title" });
+
+	const parts = memo.split("-");
+	// const transaction = await senderClient.getTransactionXdr(receiverKeys.publicKey(), amount, memo);
```
Comment (doodyparizada) [inline scripts/src/mock_client.ts]: delete comment

File: scripts/src/public/services/orders.ts [hunk 7]
```diff
@@ -550,3 +550,43 @@ export async function createOutgoingTransferOrder(recipientWalletAddress: string
...
+		meta: { title, description }
+	});
+
+	await order.save();
+
+	// we create a redis entry so this memo can uniquely identify this order.id down the line
+	// i.e. when the watch we gonna add triggers a callback
+	const parts = memo.split("-"); // extract order id from memo
+	const redis = getRedisClient();
+	await redis.async.set(`transfer:${parts[2]}`, order.id);
```
Comment (doodyparizada) [inline scripts/src/public/services/orders.ts]: Maybe extract these 3 lines to a utility function

File: scripts/src/public/services/orders.ts [hunk 8]
```diff
@@ -550,3 +550,43 @@ export async function createOutgoingTransferOrder(recipientWalletAddress: string
...
+
+	// adding a watch
+	const res = await setWatcherEndpoint([receiverWallet.address], receiver.appId);
```
Comment (doodyparizada) [inline scripts/src/public/services/orders.ts]: You shouldn't call this here. Only a call to `addWatcherEndpoint`. 
The `setWatcherEndpoint` should only be called from the initialization code of the marketplace-internal.

File: scripts/src/public/services/payment.ts [hunk 9]
```diff
@@ -123,22 +123,24 @@ export async function getPayments(walletAddress: string, options?: { timeout?: n
...
-
-	if (blockchainVersion === "3") {
-		return;
-	}
-	await httpClient.put(`${ getPaymentServiceUrl("2") }/services/${ SERVICE_ID }/watchers/${ address }/payments/${ paymentId }`);
+	logger().info("watch url will be " + `${ getPaymentServiceUrl(blockchainVersion) }/services/${ SERVICE_ID }/watchers/${ address }/payments/${ paymentId }`);
+	// if (blockchainVersion === "3") {
+	// 	return;
+	// }
+	// await httpClient.put(`${ getPaymentServiceUrl("2") }/services/${ SERVICE_ID }/watchers/${ address }/payments/${ paymentId }`);
```
Comment (doodyparizada) [inline scripts/src/public/services/payment.ts]: remove above debugging code and comments

File: scripts/src/tests/services/orders.spec.ts [hunk 10]
```diff
@@ -725,4 +726,19 @@ describe("test v2 orders", async () => {
 		const order = await createOutgoingTransferOrder(receiverWalletAddress, receiverApp.id, "a title", "a description", "a memo" , 1000, sender, deviceId1);
 		expect(order).toMatchObject({ amount: 1000 });
 	});
+
+	test("create an incoming transfer order shaharsol", async () => {
```
Comment (doodyparizada) [inline scripts/src/tests/services/orders.spec.ts]: delete `shaharsol`

File: scripts/src/client.ts [hunk 11]
```diff
@@ -417,6 +438,24 @@ export class Client {
 		}
 	}
 
+	public async createIncomingTransferOrder(walletAddress: string, appId: string, title: string, description: string, memo: string): Promise<OpenOrder> {
```
Comment (doodyparizada) [inline scripts/src/client.ts]: The return from incoming/orders is a regualr `Order`, not `OpenOrder`

File: scripts/src/internal/services.ts [hunk 12]
```diff
@@ -101,7 +102,12 @@ async function getPaymentJWT(order: db.Order, appId: string, user: User): Promis
 }
 
 export async function paymentComplete(payment: CompletedPayment) {
-	const order = await db.Order.getOne({ orderId: payment.id });
+	// if I can find the "paytment.id" in the redis, it means its a memo and the value is the real order id
+	const redis = getRedisClient();
+	const incomingOrderId = await redis.async.get(`transfer:${payment.id}`);
```
Comment (doodyparizada) [inline scripts/src/internal/services.ts]: Maybe put this into a utility function - i.e. hide redis impl

**Review by doodyparizada — CHANGES_REQUESTED** [review]
Some changes

File: scripts/src/mock_client.ts (hunk 4 above)
Comment (shaharsol) [reply scripts/src/mock_client.ts]: done

File: scripts/src/mock_client.ts (hunk 5 above)
Comment (shaharsol) [reply scripts/src/mock_client.ts]: done

File: scripts/src/mock_client.ts (hunk 6 above)
Comment (shaharsol) [reply scripts/src/mock_client.ts]: done

File: scripts/src/public/services/orders.ts (hunk 7 above)
Comment (shaharsol) [reply scripts/src/public/services/orders.ts]: typescript ftw

File: scripts/src/public/services/payment.ts (hunk 9 above)
Comment (shaharsol) [reply scripts/src/public/services/payment.ts]: done

File: scripts/src/tests/services/orders.spec.ts (hunk 10 above)
Comment (shaharsol) [reply scripts/src/tests/services/orders.spec.ts]: i.e. harakiri? done.

File: scripts/src/public/services/orders.ts (hunk 8 above)
Comment (shaharsol) [reply scripts/src/public/services/orders.ts]: elaborate on this one. I see it's called once from `initPaymentCallbacks` that is invoked on app startup, but I don't know the address to watch then. It is created dynamically.

File: scripts/src/internal/services.ts (hunk 12 above)
Comment (shaharsol) [reply scripts/src/internal/services.ts]: Feels like an extra redundant layer. What will I name it? cache? and it'll be 1-1 redis, hard to see the point.

File: scripts/src/client.ts (hunk 11 above)
Comment (shaharsol) [reply scripts/src/client.ts]: done.

File: scripts/src/public/services/orders.ts (hunk 8 above)
Comment (doodyparizada) [reply scripts/src/public/services/orders.ts]: 1. The addresses passed in `setWatcherEndpoint` are supposed to be permanent address that you will always watch.
2. calling `setWatcherEndpoint` creates or overrides these permanent addresses and and sets the webhook to cal back when payments are found. 
3. When you want to watch an address for a short period of time like in p2p or in this incoming transfer, use `addWatcherEndpoint`. I guess naming could be better.

File: scripts/src/internal/services.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (liorama) [reply scripts/src/internal/services.ts]: A nice way to create a template builder for this is to put this at the top of the file or something:
```
transferKey = (strs, v) => { return `transfer:${v}` }
```
and then use it this way here and anywhere else you need this key:
```
transferKey`${payment.id}`
```

File: scripts/src/utils/utils.ts [hunk 13]
```diff
@@ -242,3 +242,18 @@ export function cached(cache: RedisAsyncClient, keygen: (...args: any[]) => stri
...
 }
+
+export type Memo = {
+	version: string;
+	appId: string;
+	orderId: string;
+};
+
+export function parseMemo(memo: string): Memo{
+	const parts = memo.split("-");
```
Comment (liorama) [inline scripts/src/utils/utils.ts]: maybe add a comment with an example memo, it's usually easier to comprehend a visual. Not super important though.

File: scripts/src/internal/services.ts [hunk 14]
```diff
@@ -101,7 +102,12 @@ async function getPaymentJWT(order: db.Order, appId: string, user: User): Promis
 }
 
 export async function paymentComplete(payment: CompletedPayment) {
-	const order = await db.Order.getOne({ orderId: payment.id });
+	// if I can find the "paytment.id" in the redis, it means its a memo and the value is the real order id
```
[the remark this answers is not in the document]
Comment (shaharsol) [reply scripts/src/internal/services.ts]: fixed

File: scripts/src/public/routes/index.ts [hunk 15]
```diff
@@ -46,6 +47,7 @@ export function createRoutes(app: express.Express, pathPrefix?: string) {
 
 	app.post(prefix("offers/external/orders"), authenticateUser, createExternalOrder);
 	app.post(prefix("transfers/outgoing/orders"), authenticateUser, createOutgoingTransferOrder);
+	app.post(prefix("transfers/incoming/orders"), authenticateUser, createIncomingTransferOrder);
```
[the remark this answers is not in the document]
Comment (shaharsol) [reply scripts/src/public/routes/index.ts]: done

File: scripts/src/public/services/orders.ts [hunk 16]
```diff
@@ -550,3 +550,43 @@ export async function createOutgoingTransferOrder(recipientWalletAddress: string
...
+		user: receiver,
+		type: "earn",
+		wallet: receiverWallet.address,
+		meta: { title, description }
+	});
+
+	await order.save();
+
+	// we create a redis entry so this memo can uniquely identify this order.id down the line
+	// i.e. when the watch we gonna add triggers a callback
```
[the remark this answers is not in the document]
Comment (shaharsol) [reply scripts/src/public/services/orders.ts]: i won't stop

File: scripts/src/public/services/orders.ts [hunk 17]
```diff
@@ -550,3 +550,43 @@ export async function createOutgoingTransferOrder(recipientWalletAddress: string
...
+	const parts = memo.split("-"); // extract order id from memo
```
[the remark this answers is not in the document]
Comment (shaharsol) [reply scripts/src/public/services/orders.ts]: irrelevant once parseMemo is used

File: scripts/src/internal/services.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (shaharsol) [reply scripts/src/internal/services.ts]: done

File: scripts/src/utils/utils.ts (hunk 13 above)
Comment (shaharsol) [reply scripts/src/utils/utils.ts]: done

File: scripts/src/public/services/orders.ts (hunk 8 above)
Comment (shaharsol) [reply scripts/src/public/services/orders.ts]: done

**Pull request merged**